cparser/Elab.ml

```diff
@@ -1853,50 +1900,60 @@ let elab_expr vararg loc env a =
             EBinop(op, b1, b2, TPtr(ty1, []))
         | TPtr(ty1, _), TPtr(ty2, _) ->
             if not (compatible_types AttrIgnoreAll env ty1 ty2) then
-              warning "comparison between incompatible pointer types";
+              warning Compare_distinct_pointer_types "comparison of distinct pointer types (%a and %a)"
+                (print_typ env) b1.etyp (print_typ env) b2.etyp;
             EBinop(op, b1, b2, TPtr(ty1, []))
         | TPtr _, (TInt _ | TEnum _)
         | (TInt _ | TEnum _), TPtr _ ->
-            warning "comparison between integer and pointer";
+            warning Unnamed "ordered comparison between pointer and integer (%a and %a)"
+              (print_typ env) b1.etyp (print_typ env) b2.etyp;
             EBinop(op, b1, b2, TPtr(TVoid [], []))
-        | _, _ ->
+        | ty1, ty2 ->
             error "illegal comparison between types@ %a@ and %a"
-                  Cprint.typ b1.etyp Cprint.typ b2.etyp in
+              (print_typ env) b1.etyp (print_typ env) b2.etyp; in
       { edesc = resdesc; etyp = TInt(IInt, []) },env
 
 (* Elaboration of && and || *)
   and elab_logical_operator msg op a1 a2 =
-      let b1,env = elab env a1 in
-      if not (is_scalar_type env b1.etyp) then
-        err "the first argument of '%s' is not a scalar type" msg;
-      let b2,env = elab env a2 in
-      if not (is_scalar_type env b2.etyp) then
-        err "the second argument of '%s' is not a scalar type" msg;
-      { edesc = EBinop(op, b1, b2, TInt(IInt, [])); etyp = TInt(IInt, []) },env
+    let b1,env = elab env a1 in
+    let b2,env = elab env a2 in
+    if not ((is_scalar_type env b1.etyp) && (is_scalar_type env b2.etyp)) then
+      error "invalid operands to binary expression (%a and %a)"
```

Comment:
Loss of context (which binary operator?)  Plus, original messages were more precise.